src/cuckoocache.h

```diff
@@ -0,0 +1,452 @@
+                 hash_function.template operator()<4>(e) & hash_mask,
+                 hash_function.template operator()<5>(e) & hash_mask,
+                 hash_function.template operator()<6>(e) & hash_mask,
+                 hash_function.template operator()<7>(e) & hash_mask}};
+    }
+
+    /* end
+     * @returns a constexpr index that can never be inserted to */
+    constexpr uint32_t invalid() const
+    {
+        return ~(uint32_t)1;
+    }
+
+    /** allow_erase marks the element at index n as discardable. Threadsafe
+     * without any concurrent insert.
+     * @param n the index to allow erasure of
+     */
+    inline void allow_erase(uint32_t n) const
+    {
+        collection_flags.bit_set(n);
+    }
+
+    /** please_keep marks the element at index n as an entry that should be kept.
+     * Threadsafe without any concurrent insert.
+     * @param n the index to prioritize keeping
+     */
+    inline void please_keep(uint32_t n) const
+    {
+        collection_flags.bit_unset(n);
+    }
+
+    /** epoch_check handles the changing of epochs for elements stored in the
+     * cache. epoch_check should be run before every insert.
+     *
+     * First, epoch_check decrements and checks the cheap heuristic, and then does
+     * a more expensive scan if the cheap heuristic runs out. If the expensive
+     * scan suceeds, the epochs are aged and old elements are allow_erased. The
+     * cheap heuristic is reset to retrigger after the worst case growth of the
+     * current epoch's elements would exceed the epoch_size.
+     */
+    void epoch_check()
+    {
+        if ((--epoch_heuristic_counter) != 0)
+            return;
+        // count the number of elements from the latest epoch which
+        // have not been erased.
+        uint32_t epoch_unused_count = 0;
+        for (uint32_t i = 0; i < size; ++i)
+            epoch_unused_count += epoch_flags[i] &&
+                                  !collection_flags.bit_is_set(i);
+        // If there are more non-deleted entries in the current epoch than the
+        // epoch size, then allow_erase on all elements in the old epoch (marked
+        // false) and move all elements in the current epoch to the old epoch
+        // but do not call allow_erase on their indices.
+        if (epoch_unused_count >= epoch_size) {
+            for (uint32_t i = 0; i < size; ++i)
+                if (epoch_flags[i])
+                    epoch_flags[i] = false;
+                else
+                    allow_erase(i);
+            epoch_heuristic_counter = epoch_size;
+        } else
+            // reset the epoch_heuristic_counter to next do a scan when worst
+            // case behavior (no intermittent erases) would exceed epoch size,
+            // with a reasonable minimum scan size.
+            epoch_heuristic_counter = std::max(1u, std::max(epoch_size / 16,
+                        epoch_size - std::min(epoch_size, epoch_unused_count)));
```

Comment:
The std::min should be a NOP. we're in an else(epoch_size > epoch_unused_count).